File: core/src/main/java/com/vzome/core/kinds/AbstractSymmetryPerspective.java
package com.vzome.core.kinds;

import java.util.ArrayList;
import java.util.List;

import com.vzome.core.algebra.AlgebraicNumber;
import com.vzome.core.commands.Command;
import com.vzome.core.commands.CommandSymmetry;
import com.vzome.core.editor.SymmetryPerspective;
import com.vzome.core.editor.api.Shapes;
import com.vzome.core.math.symmetry.Axis;
import com.vzome.core.math.symmetry.Direction;
import com.vzome.core.math.symmetry.OctahedralSymmetry;
import com.vzome.core.math.symmetry.Symmetry;

public abstract class AbstractSymmetryPerspective implements SymmetryPerspective
{
    protected final Symmetry symmetry;

    private final List<Shapes> geometries = new ArrayList<>();

    private Shapes defaultShapes = null;

    public AbstractSymmetryPerspective(Symmetry symmetry) {
        this.symmetry = symmetry;
    }

    @Override
    public Symmetry getSymmetry() {
        return this.symmetry;
    }

    @Override
    public String getName() {
        return getSymmetry().getName();
    }

    @Override
    public String getLabel() {
        return getSymmetry().getName();
    }

    protected void addShapes(Shapes shapes) {
        Shapes old = getGeometry(shapes.getName());
        if(old != null) {
            this.geometries.remove(old);
        }
        this.geometries.add(shapes);
    }

    protected void clearShapes() {
        this.geometries.clear();
        defaultShapes= null;
    }

    @Override
    public List<Shapes> getGeometries() {
        return this.geometries;
    }

    private Shapes getGeometry(String name) {
        for(Shapes shapes : geometries) {
            if(shapes.getName().equals(name)) {
                return shapes;
            }
        }
        return null;
    }

    public void setDefaultGeometry(Shapes shapes) {
        this.defaultShapes = shapes;
        addShapes(shapes);
    }

    @Override
    public Shapes getDefaultGeometry() {
        return this.defaultShapes;
    }

    @Override
    public Command getLegacyCommand( String action )
    {
        switch ( action ) {
        case "octasymm":
        {
            Symmetry octaSymm = getSymmetry();
            if(! (octaSymm instanceof OctahedralSymmetry) ) {
                // only make a new OctahedralSymmetry if necessary
                octaSymm = new OctahedralSymmetry(octaSymm.getField());
            }
            // This command will be availble to all SymmetryPerspectives even if they are not Octahedral
            // TODO: This legacy command should probably eventually be removed
            // after we ensure that doing so is backward compatible. 
            return new CommandSymmetry( octaSymm );
        }
        
        default:
            return null;
        }
    }

    @Override
    public boolean orbitIsStandard( Direction orbit )
    {
        return orbit .isStandard();
    }

    @Override
    public boolean orbitIsBuildDefault( Direction orbit )
    {
        Axis zone0 = orbit .getAxis( 0, 0 );
        return zone0 .getRotationPermutation() != null;
    }
    
    @Override
    public AlgebraicNumber getOrbitUnitLength( Direction orbit )
    {
        return orbit .getUnitLength();
    }
}

File: online/src/worker/legacy/from-java/com/vzome/core/kinds/AbstractSymmetryPerspective.ts
import { java, javaemul } from "../../../../../candies/j4ts-2.1.0-SNAPSHOT/bundle.js";
import { AlgebraicNumber } from "../algebra/AlgebraicNumber.js";
import { Command } from "../commands/Command.js";
import { CommandSymmetry } from "../commands/CommandSymmetry.js";
import { CommandTetrahedralSymmetry } from "../commands/CommandTetrahedralSymmetry.js";
import { SymmetryPerspective } from "../editor/SymmetryPerspective.js";
import { Shapes } from "../editor/api/Shapes.js";
import { Axis } from "../math/symmetry/Axis.js";
import { Direction } from "../math/symmetry/Direction.js";
import { OctahedralSymmetry } from "../math/symmetry/OctahedralSymmetry.js";
import { Symmetry } from "../math/symmetry/Symmetry.js";

export abstract class AbstractSymmetryPerspective implements SymmetryPerspective {
    symmetry: Symmetry;

    /*private*/ geometries: java.util.List<Shapes>;

    /*private*/ defaultShapes: Shapes;

    public constructor(symmetry: Symmetry) {
        if (this.symmetry === undefined) { this.symmetry = null; }
        this.geometries = <any>(new java.util.ArrayList<any>());
        this.defaultShapes = null;
        this.symmetry = symmetry;
    }

    /**
     * 
     * @return {*}
     */
    public getSymmetry(): Symmetry {
        return this.symmetry;
    }

    /**
     * 
     * @return {string}
     */
    public getName(): string {
        return this.getSymmetry().getName();
    }

    /**
     * 
     * @return {string}
     */
    public getLabel(): string {
        return this.getSymmetry().getName();
    }

    addShapes(shapes: Shapes) {
        const old: Shapes = this.getGeometry(shapes.getName());
        if (old != null){
            this.geometries.remove(old);
        }
        this.geometries.add(shapes);
    }

    clearShapes() {
        this.geometries.clear();
        this.defaultShapes = null;
    }

    /**
     * 
     * @return {*}
     */
    public getGeometries(): java.util.List<Shapes> {
        return this.geometries;
    }

    /*private*/ getGeometry(name: string): Shapes {
        for(let index=this.geometries.iterator();index.hasNext();) {
            let shapes = index.next();
            {
                if (shapes.getName() === name){
                    return shapes;
                }
            }
        }
        return null;
    }

    public setDefaultGeometry(shapes: Shapes) {
        this.defaultShapes = shapes;
        this.addShapes(shapes);
    }

    /**
     * 
     * @return {*}
     */
    public getDefaultGeometry(): Shapes {
        return this.defaultShapes;
    }

    /**
     * 
     * @param {string} action
     * @return {*}
     */
    public getLegacyCommand(action: string): Command {
        switch((action)) {
        case "octasymm":
            {
                let octaSymm: Symmetry = this.getSymmetry();
                if (!(octaSymm != null && octaSymm instanceof <any>OctahedralSymmetry)){
                    octaSymm = new OctahedralSymmetry(octaSymm.getField());
                }
                return new CommandSymmetry(octaSymm);
            };
        case "tetrasymm":
            {
                const symmetry: Symmetry = this.getSymmetry();
                return new CommandTetrahedralSymmetry(symmetry);
            };
        default:
            return null;
        }
    }

    /**
     * 
     * @param {Direction} orbit
     * @return {boolean}
     */
    public orbitIsStandard(orbit: Direction): boolean {
        return orbit.isStandard();
    }

    /**
     * 
     * @param {Direction} orbit
     * @return {boolean}
     */
    public orbitIsBuildDefault(orbit: Direction): boolean {
        const zone0: Axis = orbit.getAxis$int$int(0, 0);
        return zone0.getRotationPermutation() != null;
    }

    /**
     * 
     * @param {Direction} orbit
     * @return {*}
     */
    public getOrbitUnitLength(orbit: Direction): AlgebraicNumber {
        return orbit.getUnitLength();
    }

    public abstract createToolFactories(kind?: any, model?: any): any;
    public abstract getModelResourcePath(): any;
    public abstract predefineTools(kind?: any, model?: any): any;    }
AbstractSymmetryPerspective["__class"] = "com.vzome.core.kinds.AbstractSymmetryPerspective";
AbstractSymmetryPerspective["__interfaces"] = ["com.vzome.core.editor.SymmetryPerspective"];
